plugins/Disintegrator/Disintegrator.cpp

```diff
@@ -0,0 +1,240 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "Disintegrator.h"
+
+#include "embed.h"
+#include "lmms_math.h"
+#include "plugin_export.h"
+
+extern "C"
+{
+
+Plugin::Descriptor PLUGIN_EXPORT disintegrator_plugin_descriptor =
+{
+	STRINGIFY(PLUGIN_NAME),
+	"Disintegrator",
+	QT_TRANSLATE_NOOP("pluginBrowser", "A delay modulation effect for very aggressive digital distortion."),
+	"Lost Robot <r94231@gmail.com>",
+	0x0100,
+	Plugin::Effect,
+	new PluginPixmapLoader("logo"),
+	NULL,
+	NULL
+} ;
+
+}
+
+
+
+DisintegratorEffect::DisintegratorEffect(Model* parent, const Descriptor::SubPluginFeatures::Key* key) :
+	Effect(&disintegrator_plugin_descriptor, parent, key),
+	m_disintegratorControls(this),
+	m_needsUpdate(true)
+{
+	sampleRateChanged();
+}
+
+
+
+void DisintegratorEffect::sampleRateChanged()
+{
+	m_sampleRate = Engine::mixer()->processingSampleRate();
+	m_2PiOverSR =  F_2PI / m_sampleRate;
+	m_lp = new BasicFilters<2>(Engine::mixer()->processingSampleRate());
+	m_hp = new BasicFilters<2>(Engine::mixer()->processingSampleRate());
+	m_lp->setFilterType(0);
+	m_hp->setFilterType(1);
+	m_needsUpdate = true;
+
+	m_bufferSize = m_sampleRate * 0.01f + 1.f;
+	for (int i = 0; i < 2; ++i)
+	{
+		m_inBuf[i].resize(m_bufferSize);
+	}
+}
+
+
+
+bool DisintegratorEffect::processAudioBuffer(sampleFrame* buf, const fpp_t frames)
+{
+	if (!isEnabled() || !isRunning ())
+	{
+		return false;
+	}
+
+	double outSum = 0.0;
+	const float d = dryLevel();
+	const float w = wetLevel();
+	
+	const ValueBuffer * amountBuf = m_disintegratorControls.m_amountModel.valueBuffer();
+	const ValueBuffer * typeBuf = m_disintegratorControls.m_typeModel.valueBuffer();
+	const ValueBuffer * freqBuf = m_disintegratorControls.m_lowCutModel.valueBuffer();
+
+	// Update filters
+	if(m_needsUpdate || m_disintegratorControls.m_highCutModel.isValueChanged())
+	{
+		m_lp->calcFilterCoeffs(m_disintegratorControls.m_highCutModel.value(), 0.5);
+		
+	}
+	if(m_needsUpdate || m_disintegratorControls.m_lowCutModel.isValueChanged())
+	{
+		m_hp->calcFilterCoeffs(m_disintegratorControls.m_lowCutModel.value(), 0.5);
+	}
+	m_needsUpdate = false;
+
+	for (fpp_t f = 0; f < frames; ++f)
+	{
+		const float amount = (amountBuf ? amountBuf->value(f) : m_disintegratorControls.m_amountModel.value()) * 0.001f * m_sampleRate;
+		const int type = typeBuf ? typeBuf->value(f) : m_disintegratorControls.m_typeModel.value();
+		const float freq = freqBuf ? freqBuf->value(f) : m_disintegratorControls.m_freqModel.value();
+	
+		sample_t s[2] = {buf[f][0], buf[f][1]};
+
+		// Increment buffer read point
```

Comment:
when implementing a circular buffer, it is common to have the buffer size a power of 2, with 512 being the next closest to the 500 you stare in the header.

The index increment and wrap at end of buffer can then be reduced to: 
``` 
++m_inBufLoc;
m_inBufLoc &= (m_bufferSize - 1)
```

This is preferable as conditional branches can be processor expensive, and may also hinder some compiler side optimisations.